### ISSUE TITLE
Some policy implementation for the Globus connector client seems not possible.

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#8103]
### iRODS Version, OS and Version
4.3.3, almalinux9
iCommands 4.3.3
irods_client_globus_connector 4.3.3.0

### What did you try to do?
We need to block specific files transferred through the irods_client_globus_connector from being written to iRODS.
For this aim, we wanted to implement this PEP rule:

```
def pep_api_data_obj_open_pre(rule_args, callback, rei):
    user_name = session_vars.get_map(rei)["client_user"]["user_name"]
    object_path = object_path = str(rule_args[2].objPath)
    if object_path == "/icts/home/u0137480/myfile":
        # blocking code
        block_it
```

### Expected behavior
Expected only the file that is needed should be blocked, allowing the client to continue transferring other queued files and return a single error.

### Observed behavior (including steps to reproduce, if applicable)
Although the file is blocked, the client continuously attempts to transfer/write it, blocking other queued files.

Transfer some files:
<img width="427" alt="image" src="https://github.com/user-attachments/assets/72671c9b-a754-4099-be14-29359d3a193e" />

Appearance in irods:
```
root@d4aeffbf530b:/# ils -L /icts/home/u0137480/globus_test/test
/icts/home/u0137480/globus_test/test:
  u0137480          0 default;netapp     29360128 2024-12-18.09:13 & 50MB2.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB2.zip
root@d4aeffbf530b:/# ils -L /icts/home/u0137480/globus_test/test
/icts/home/u0137480/globus_test/test:
  u0137480          0 default;netapp     29622272 2024-12-18.09:14 & 50MB2.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB2.zip
  u0137480          0 default;netapp            0 2024-12-18.09:14 ? 50MB4.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB4.zip
/icts/home/u0137480/globus_test/test:
  u0137480          0 default;netapp     37224448 2024-12-18.09:21 & 50MB2.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB2.zip
  u0137480          0 default;netapp            0 2024-12-18.09:14 & 50MB4.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB4.zip
  u0137480          0 default;netapp            0 2024-12-18.09:16 & 50MB5.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB5.zip
```

client side:
<img width="525" alt="image" src="https://github.com/user-attachments/assets/7dabf0f8-e6f2-486c-997b-773408f4b1e4" />
<img width="150" alt="image" src="https://github.com/user-attachments/assets/b91eabea-599f-4fa1-a535-37877e67501f" />
<img width="508" alt="image" src="https://github.com/user-attachments/assets/26565b2c-c8be-4ad2-a14e-5358ba4e4616" />

irods logs:
```
[2024-12-18T09:26:56.975Z][icts-t-cloud-rdm-leu-2] {"processname": "irods", "groupname": "irods", "pid": "95", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = object path from pep_api_data_obj_open_post is: /icts/home/u0137480/globus_test/test/50MB2.zip\n", "request_api_name": "DATA_OBJ_OPEN_AN", "request_api_number": 602, "request_api_version": "d", "request_client_user": "u0137480", "request_host": "127.0.0.1", "request_proxy_user": "globus", "request_release_version": "rods4.3.3", "server_host": "icts.irods.t.icts.kuleuven.be", "server_pid": 8128, "server_timestamp": "2024-12-18T09:26:56.974Z", "server_type": "agent", "server_zone": "icts"}}
[2024-12-18T09:26:57.758Z][icts-t-cloud-rdm-leu-2] {"processname": "irods", "groupname": "irods", "pid": "95", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "USER_NOT_ALLOWED_TO_EXEC_CMD: Wrong files cannot be uploaded!", "request_api_name": "DATA_OBJ_OPEN_AN", "request_api_number": 602, "request_api_version": "d", "request_client_user": "u0137480", "request_host": "127.0.0.1", "request_proxy_user": "globus", "request_release_version": "rods4.3.3", "server_host": "icts.irods.t.icts.kuleuven.be", "server_pid": 8159, "server_timestamp": "2024-12-18T09:26:57.757Z", "server_type": "agent", "server_zone": "icts"}}
```

In this example, the file "/icts/home/u0137480/globus_test/test/50MB3.zip" is blocked by policy. However, this blocking behavior is causing unintended consequences. All other files in the transfer queue are stuck (hanging) in a partially written state, as shown above.

How can we implement a policy that will block some certain files?

**Comment by trel** [comment]
The server doesn't know that it's Globus making a request, specifically.

The policy would have to block based on some other criteria (username, client IP, time of day, logical path... etc.)

And separately... whether the client is 'retrying' and/or blocking other transfers is out of scope for the iRODS server.

I will note that the `pep_api_data_obj_open_pre` PEP will also block reads.

**Comment by mstfdkmn** [comment]
> The policy would have to block based on some other criteria (username, client IP, time of day, logical path... etc.)

Yes, we have additional criteria/checks. As shown in the example above, I added only a specific logical path for control. Therefore, I ca say we only want to block certain files transferred through Globus to certain collections.

> And separately... whether the client is 'retrying' and/or blocking other transfers is out of scope for the iRODS server.

You're right. And I created this issue here rather than in the Globus connector repository because it concerns policy implementation in general/it's a broader issue about how to implement blocking policies. I'm wondering whether the server can provide a response to the globus client when a blocking rule is triggered so that the client can continue with other files, or if the Globus connector can be designed to skip blocked files.

We welcome any alternative approaches. To clarify our need: when transferring a data set containing multiple files via globus, we would like to block files with lets say the .txt extension.

**Comment by trel** [comment]
Globus policy on retry will control whether the same file is attempted multiple times.  Are you saying that by blocking one file, the other files in the Globus 'queue' don't get sent?  Or just that they are delayed while they wait for the blocked file to fail a certain number of times?

I'm confused by the zero-length files in iRODS.   That is not what I would have expected to see.   @alanking is that due to an order-of-operations in the flow of PEPs?

**Comment by mstfdkmn** [comment]
> Are you saying that by blocking one file, the other files in the Globus 'queue' don't get sent?

Yes, partially. As shown in the screenshot of the target collection, unblocked files are being (as if) written but are not completing. The `ils` output shows one partially written file and two zero-size files that remain incomplete. However the client side report (0 B
Bytes Transferred) shows no file is being written.

**Comment by alanking** [comment]
I'm not sure whether I have any good input for the policy implementation, but one thing stuck out to me when looking at that `ils` output...

It appears that the Globus connector is still opening/closing data objects while data is still in flight. Notice that 50MB2.zip is growing in size even though it is marked as at rest. There was an issue in earlier versions of the Globus connector around opening/closing files as data is being transferred, but I guess that has been closed. Here's that issue: https://github.com/irods/irods_client_globus_connector/issues/84

I am wondering whether this has any effect on the Globus view into whether something is "complete" or not. I don't know enough about that tool to say either way, but it appears that the object is being finalized multiple times over the course of its transfer, which feels like an incorrect implementation of a parallel transfer into iRODS.

**Comment by mstfdkmn** [comment]
> It appears that the Globus connector is still opening/closing data objects while data is still in flight. Notice that 50MB2.zip is growing in size even though it is marked as at rest. 

The size reported by ils for the file "50MB2.zip" fluctuates (increasing/decreasing) during the transfer, making it difficult to assess progress accurately. The provided additional ils output for the same transfer below confirms this inconsistency. Furthermore, the transfer logic seems working with alphanumeric order. In 12 minutes, it has only attempted to open "50MB2.zip, 50MB4.zip, 50MB5.zip," and hasn't even begun processing "50MB6.zip."

```
/icts/home/u0137480/globus_test/test:
  u0137480          0 default;netapp     27787264 2024-12-18.09:32 & 50MB2.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB2.zip
  u0137480          0 default;netapp            0 2024-12-18.09:14 & 50MB4.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB4.zip
  u0137480          0 default;netapp            0 2024-12-18.09:16 & 50MB5.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB5.zip
root@d4aeffbf530b:/# ils -L /icts/home/u0137480/globus_test/test
/icts/home/u0137480/globus_test/test:
  u0137480          0 default;netapp     28573696 2024-12-18.09:37 & 50MB2.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB2.zip
  u0137480          0 default;netapp            0 2024-12-18.09:14 & 50MB4.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB4.zip
  u0137480          0 default;netapp            0 2024-12-18.09:16 & 50MB5.zip
        generic    /netapp/home/u0137480/globus_test/test/50MB5.zip
```

> I am wondering whether this has any effect on the Globus view into whether something is "complete" or not. 

We also have limited information, but it appears that the Globus interface/client is only updated upon successful file completion or when no error occurs.

**Comment by trel** [comment]
I don't think there is much for us to do in the server at this time for this issue.   How Globus handles its writes and its queuing is out of scope here.

Moving this issue to the connector repository may be most appropriate.

**Comment by korydraughn** [comment]
I agree. The issue is out of scope for the iRODS server.

The coordination needs to occur between the connector and globus.
